File: java/backup-to-storage/src/main/java/BackupToStorage.java
import com.mashape.unirest.http.HttpResponse;
import com.mashape.unirest.http.JsonNode;
import com.mashape.unirest.http.Unirest;
import io.appwrite.exceptions.AppwriteException;
import org.json.JSONArray;
import org.json.JSONObject;
import java.io.FileWriter;
import java.io.IOException;

class BackupToStorage {
    public static void main(String[] args) throws IOException, AppwriteException {
        FileWriter collectionsCsv = null;
        String headers = "collection_id, permissions, name, date_created, date_updated, rules";
        try {
            String filePath = System.getProperty("user.dir") + "/collections.csv";
            collectionsCsv = new FileWriter(filePath);

            collectionsCsv.append(headers);
            collectionsCsv.append('\n');

            HttpResponse<JsonNode> response =
                    Unirest.get(String.format("%s/database/collections", System.getenv("APPWRITE_ENDPOINT")))
                            .header("X-Appwrite-Project", System.getenv("APPWRITE_FUNCTION_PROJECT_ID"))
                            .header("X-Appwrite-key", System.getenv("APPWRITE_API_KEY"))
                            .asJson();


            JSONObject responeJson = response.getBody().getObject();
            JSONArray collections = responeJson.getJSONArray("collections");

            for (int i=0; i < collections.length(); i++) {
                JSONObject jsonCollection = collections.getJSONObject(i);

                collectionsCsv.append(jsonCollection.getString("$id"));
                collectionsCsv.append(",");
                collectionsCsv.append(jsonCollection.getJSONObject("$permissions").toString());
                collectionsCsv.append(",");
                collectionsCsv.append(jsonCollection.getString("name"));
                collectionsCsv.append(",");
                collectionsCsv.append((char) jsonCollection.getLong("dateCreated"));
                collectionsCsv.append(",");
                collectionsCsv.append((char) jsonCollection.getLong("dateUpdated"));
                collectionsCsv.append(",");
                collectionsCsv.append(jsonCollection.getJSONArray("rules").toString());
                collectionsCsv.append("\n");
            }

            HttpResponse<JsonNode> storageResponse =
                    Unirest.get(String.format("%s/storage/files", System.getenv("APPWRITE_ENDPOINT")))
                            .header("X-Appwrite-Project", System.getenv("APPWRITE_FUNCTION_PROJECT_ID"))
                            .header("X-Appwrite-key", System.getenv("APPWRITE_API_KEY"))
                            .asJson();


            if (storageResponse.getStatus() >= 200) {
                System.out.println("Successfully uploaded collections.csv to storage");
            }

        } catch (Throwable th) {
            System.out.println(th);
        } finally {
            if (collectionsCsv != null) {
                collectionsCsv.flush();
            }
            System.exit(0);
        }
    }
}

File: kotlin/backup-to-storage/src/main/kotlin/BackupToStorage.kt
import com.google.gson.Gson
import io.appwrite.Client
import io.appwrite.services.Database
import io.appwrite.services.Storage
import kotlinx.coroutines.runBlocking
import java.io.File
import java.io.FileWriter
import kotlin.system.exitProcess


class BackupToStorage {
    companion object {
        private var collectionsCsv: FileWriter? = null
        private const val headers = "collection_id, permissions, name, date_created, date_updated, rules"

        @JvmStatic
        fun main(args: Array<String>) {
            try {
                val appWriteclient = Client()
                    .setEndpoint(System.getenv("APPWRITE_ENDPOINT"))
                    .setProject(System.getenv("APPWRITE_FUNCTION_PROJECT_ID"))
                    .setKey(System.getenv("APPWRITE_API_KEY"))


                collectionsCsv = FileWriter("${System.getProperty("user.dir")}/collections.csv")
                if (collectionsCsv == null) {
                    throw Exception("Error in creating file")
                }

                collectionsCsv!!.append(headers)
                collectionsCsv!!.append('\n')

                val database = Database(appWriteclient)
                val response = runBlocking {
                    database.listCollections().body?.string()
                }

                val responseBody = Gson().fromJson(response, Collections::class.java) ?: throw Exception("Empty response")

                for (collection in responseBody.collections) {
                    collectionsCsv!!.append(collection.id)
                    collectionsCsv!!.append(",")
                    if (collection.permissions != null) {
                        collectionsCsv!!.append(collection.permissions.toString())
                    } else {
                        collectionsCsv!!.append("")
                    }
                    collectionsCsv!!.append(",")
                    collectionsCsv!!.append(collection.name)
                    collectionsCsv!!.append(",")
                    collectionsCsv!!.append(collection.dateCreated.toString())
                    collectionsCsv!!.append(",")
                    collectionsCsv!!.append(collection.dateUpdated.toString())
                    collectionsCsv!!.append(",")
                    collectionsCsv!!.append(collection.rules.toString())
                    collectionsCsv!!.append("\n")
                }
                val storage = Storage(appWriteclient)

                val storageResponse = runBlocking {
                    storage.createFile(File("${System.getProperty("user.dir")}/collections.csv"))
                }

                if (storageResponse.code >= 200) {
                    println("Successfully uploaded collections.csv to storage")
                }
            } catch (e: Exception) {
                println(e)
            } finally {
                if (collectionsCsv != null) {
                    collectionsCsv!!.flush()
                }
                exitProcess(0)
            }
        }
    }
}
